### ISSUE TITLE
Add support for Per-Sample Adaptive Clipping (PSAC)

### ISSUE THREAD

**Issue by ParthS007** [meta-pytorch/opacus#799]
## 🚀 Feature

Add Per-Sample Adaptive Clipping (PSAC) support to Opacus with PSACDPOptimizer that use a non-monotonic adaptive weight function for per-sample clipping thresholds, reducing deviation from the true batch gradient.

## Motivation

Manual tuning of the clipping threshold is time-consuming and costly. Normalization-based methods (Auto-S, NSGD) use a monotonic weight function that over-weights small gradients, increasing deviation from the true batch-averaged gradient. PSAC addresses this with a non-monotonic adaptive weight function that gives similar-order weights to samples with different gradient norms.

## Pitch

Implement PSAC based on "Differentially Private Learning with Per-Sample Adaptive Clipping": https://arxiv.org/pdf/2212.00328
- Per-sample adaptive clipping thresholds (not a single global norm)
- Non-monotonic weight function to avoid over-weighting small gradients
- Reduces deviation between clipped and true batch gradients compared to normalization-based approaches

## Usage

```
privacy_engine = PrivacyEngine()
model, optimizer, dataloader = privacy_engine.make_private(
    module=model,
    optimizer=optimizer,
    data_loader=dataloader,
    noise_multiplier=1.0,
    max_grad_norm=1.0,
    clipping="psac",  # New option
    r=0.01,           # Optional: weight function parameter
    tau0=0.1,         # Optional: lower threshold
    tau1=0.5,         # Optional: upper threshold
)
```

- Better gradient approximation than normalization-based methods
- Reduces hyperparameter tuning burden compared to constant clipping

## Alternatives

- Existing `AdaClipDPOptimizer`: adapts a single global clipping norm over time, not per-sample.
- Constant clipping (`DPOptimizer`): requires manual threshold tuning

## Additional context

I'm happy to submit a PR for this feature if it aligns with the project's goals. I'm implementing Per-Sample Adaptive Clipping as part of my thesis work, and I believe making it available in the main library could benefit the broader Opacus community.

**Comment by ParthS007** [comment]
cc: @HuanyuZhang @EnayatUllah :)
